File: python/compresr/retry.py
"""Bounded retry policy for the SDK transport layer.

Defaults retry ``429`` and ``503`` from the proxy with exponential backoff +
jitter. Pass ``RetryConfig(max_retries=0)`` to opt out.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import Tuple


@dataclass(frozen=True)
class RetryConfig:
    """Retry policy for transient backpressure responses.

    Args:
        max_retries: Maximum number of retry attempts after the initial
            request fails with a retryable status. ``0`` disables retries.
        initial_backoff_s: Backoff before the first retry, in seconds.
        max_backoff_s: Upper bound for any single backoff.
        multiplier: Exponential growth factor between attempts.
        jitter: Fractional symmetric jitter (e.g. ``0.25`` means the actual
            sleep is uniformly sampled from ``[delay * 0.75, delay * 1.25]``).
            Set to ``0`` for deterministic backoff in tests.
        retry_on_status: HTTP status codes that trigger a retry. Defaults
            to ``(429, 503)`` — the two backpressure signals the proxy uses.
        respect_retry_after: When the server returns a ``retry_after`` hint
            (header or body field), use it instead of the computed backoff
            for that attempt.
    """

    max_retries: int = 3
    initial_backoff_s: float = 0.5
    max_backoff_s: float = 30.0
    multiplier: float = 2.0
    jitter: float = 0.25
    retry_on_status: Tuple[int, ...] = field(default_factory=lambda: (429, 503))
    respect_retry_after: bool = True

    def __post_init__(self) -> None:
        if self.max_retries < 0:
            raise ValueError("max_retries must be >= 0")
        if self.initial_backoff_s < 0:
            raise ValueError("initial_backoff_s must be >= 0")
        if self.max_backoff_s < 0:
            raise ValueError("max_backoff_s must be >= 0")
        if self.multiplier < 1:
            raise ValueError("multiplier must be >= 1")
        if not 0 <= self.jitter <= 1:
            raise ValueError("jitter must be in [0, 1]")


def compute_backoff(
    attempt: int,
    cfg: RetryConfig,
    retry_after: float | None = None,
) -> float:
    """Compute the sleep duration before retry ``attempt`` (0-indexed).

    Honors ``retry_after`` from the server when present and
    ``cfg.respect_retry_after`` is true; otherwise applies exponential
    backoff with optional symmetric jitter, capped at ``max_backoff_s``.
    """
    if retry_after is not None and cfg.respect_retry_after:
        # Server-suggested delay still respects the global cap so a buggy
        # backend can't pin the client into a multi-minute sleep.
        return max(0.0, min(float(retry_after), cfg.max_backoff_s))

    base = cfg.initial_backoff_s * (cfg.multiplier**attempt)
    base = min(base, cfg.max_backoff_s)
    if cfg.jitter == 0:
        return base
    spread = base * cfg.jitter
    return max(0.0, base + random.uniform(-spread, spread))

File: typescript/src/http/retry.ts
export interface RetryConfig {
  maxRetries?: number;
  initialBackoffMs?: number;
  maxBackoffMs?: number;
  multiplier?: number;
  /** Fractional symmetric jitter in `[0, 1]` (e.g. `0.25` = ±25%). */
  jitter?: number;
  retryOnStatus?: number[];
  respectRetryAfter?: boolean;
}

export interface ResolvedRetryConfig {
  maxRetries: number;
  initialBackoffMs: number;
  maxBackoffMs: number;
  multiplier: number;
  jitter: number;
  retryOnStatus: ReadonlySet<number>;
  respectRetryAfter: boolean;
}

export const DEFAULT_RETRY_CONFIG: ResolvedRetryConfig = {
  maxRetries: 3,
  initialBackoffMs: 500,
  maxBackoffMs: 30_000,
  multiplier: 2.0,
  jitter: 0.25,
  retryOnStatus: new Set<number>([429, 503]),
  respectRetryAfter: true,
};

export function resolveRetryConfig(cfg?: RetryConfig): ResolvedRetryConfig {
  if (!cfg) return DEFAULT_RETRY_CONFIG;
  const merged: ResolvedRetryConfig = {
    maxRetries: cfg.maxRetries ?? DEFAULT_RETRY_CONFIG.maxRetries,
    initialBackoffMs: cfg.initialBackoffMs ?? DEFAULT_RETRY_CONFIG.initialBackoffMs,
    maxBackoffMs: cfg.maxBackoffMs ?? DEFAULT_RETRY_CONFIG.maxBackoffMs,
    multiplier: cfg.multiplier ?? DEFAULT_RETRY_CONFIG.multiplier,
    jitter: cfg.jitter ?? DEFAULT_RETRY_CONFIG.jitter,
    retryOnStatus: cfg.retryOnStatus
      ? new Set<number>(cfg.retryOnStatus)
      : DEFAULT_RETRY_CONFIG.retryOnStatus,
    respectRetryAfter: cfg.respectRetryAfter ?? DEFAULT_RETRY_CONFIG.respectRetryAfter,
  };
  if (merged.maxRetries < 0) throw new RangeError('maxRetries must be >= 0');
  if (merged.initialBackoffMs < 0) throw new RangeError('initialBackoffMs must be >= 0');
  if (merged.maxBackoffMs < 0) throw new RangeError('maxBackoffMs must be >= 0');
  if (merged.multiplier < 1) throw new RangeError('multiplier must be >= 1');
  if (merged.jitter < 0 || merged.jitter > 1) {
    throw new RangeError('jitter must be in [0, 1]');
  }
  return merged;
}

export function computeBackoffMs(
  attempt: number,
  cfg: ResolvedRetryConfig,
  retryAfterSeconds?: number
): number {
  if (retryAfterSeconds !== undefined && cfg.respectRetryAfter) {
    const ms = Math.max(0, retryAfterSeconds * 1000);
    return Math.min(ms, cfg.maxBackoffMs);
  }
  const raw = cfg.initialBackoffMs * Math.pow(cfg.multiplier, attempt);
  const capped = Math.min(raw, cfg.maxBackoffMs);
  if (cfg.jitter === 0) return capped;
  const spread = capped * cfg.jitter;
  return Math.max(0, capped + (Math.random() * 2 - 1) * spread);
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
